### PR TITLE
DTag transfer implementation

### LINKED ISSUE #225: Allows the users to give their tag away
## Feature description 
Currently when the user creates a profile, his DTag is selected once initially and then it can never change. However, I think there might be some cases in which two users might agree on a DTag exchange, or one user might want to give away his DTag to another user. 

An example could be a user registering the DTag `microsoft` initially, and then wanting to sell that DTag to the real Microsoft company later.

## Implementation proposal
In order to implement this possibility, I think we can simply create a new message type, named maybe something like `MsgTransferDTag` formed as follows:

```golang
type MsgTransferDTag struct {
  CurrentOwner  sdk.AccAddress `json:"current_owner" yaml:"current_owner"`
  ReceivingUser sdk.AccAddress `json:"receiving_user" yaml:"receiving_user"`
}
```

Thsi message, when handled, would simply take the current DTag of the `CurrentOwner` and transfer it to the `ReceivingUser`. Also, if the `ReceivingUser` already has a DTag, such DTag will be transfered to the `CurrentOwner`, de-facto creating a mechanism by which two users can exchange their tags. 

What do you guys think about this? @kwunyeung @bragaz

### PR DESCRIPTION
**Pull request by bragaz**
## Description
This PR implements the possibility to exchange/trade a `profile` `dTag`.
Closes #225.
## Checklist
- [x] Targeted PR against correct branch.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit tests.
- [x] Wrote integration tests (simulation & CLI). 
- [x] Updated the documentation. 
- [x] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: docs/developers/msgs/profiles/accept-dtag-transfer-request.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# `MsgAcceptDTagTransfer`
+This message allows you to accept a `dTag`'s transfer request.
```
Comment (RiccardoM) [inline docs/developers/msgs/profiles/accept-dtag-transfer-request.md:2]:
```suggestion
This message allows you to accept a `DTag`'s transfer request.
```

File: docs/developers/msgs/relationships/create-relationship.md [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # `MsgCreateRelationship`
-This message allows you to create a relationship between the signer and a specified user.
+This message allows you to create a relationship between the signer, and a specified user.
```
Comment (RiccardoM) [inline docs/developers/msgs/relationships/create-relationship.md:2]:
```suggestion
This message allows you to create a relationship between the signer and a specified user.
```

File: docs/developers/queries/user-dtag-requests.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+## Query user's dTag requests
```
Comment (RiccardoM) [inline docs/developers/queries/user-dtag-requests.md:1]:
```suggestion
## Query user's DTag requests
```

File: docs/developers/queries/user-dtag-requests.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+This query endpoint allows you to retrieve the dTag requests of the user with the given `address`.
```
Comment (RiccardoM) [inline docs/developers/queries/user-dtag-requests.md:2]:
```suggestion
This query endpoint allows you to retrieve the DTag requests of the user with the given `address`.
```

File: docs/developers/query-data.md [hunk 5]
```diff
@@ -14,6 +14,7 @@ Aside from [performing transactions](perform-transactions.md) you can also query
 ## Profiles
 - [Query a profile](queries/profile.md)
 - [Query the stored profiles](queries/profiles.md)
+- [Query all the dTag transfer's requests](queries/user-dtag-requests.md)
```
Comment (RiccardoM) [inline docs/developers/query-data.md:17]:
```suggestion
- [Query all the DTag transfer's requests](queries/user-dtag-requests.md)
```

File: docs/types/profiles/dtag-transfer-request.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# DTag transfer request
+DTag transfer request represents the request made from a user to get the dTag of another one.
```
Comment (RiccardoM) [inline docs/types/profiles/dtag-transfer-request.md:2]:
```suggestion
DTag transfer request represents the request made from a user to get the DTag of another one.
```

File: x/profiles/client/rest/rest.go [hunk 7]
```diff
@@ -26,3 +26,14 @@ type SaveProfileReq struct {
...
+
+type TransferDTagReq struct {
+	BaseReq      rest.BaseReq `json:"base_req"`
+	CurrentOwner string       `json:"current_owner"`
+}
+
+type AcceptDTagTransferReq struct {
+	BaseReq       rest.BaseReq `json:"base_req"`
+	ReceivingUser string       `json:"receiving_user"`
+	NewDTag       string       `json:"new_d_tag"`
```
Comment (RiccardoM) [inline x/profiles/client/rest/rest.go:38]: Can we rename this parameter to `new_dtag` instead?

```suggestion
	NewDTag       string       `json:"new_dtag"`
```

File: x/profiles/simulation/decoder.go [hunk 8]
```diff
@@ -23,6 +23,11 @@ func DecodeStore(cdc *codec.Codec, kvA, kvB kv.Pair) string {
 		cdc.MustUnmarshalBinaryBare(kvA.Value, &addressA)
 		cdc.MustUnmarshalBinaryBare(kvB.Value, &addressB)
 		return fmt.Sprintf("AddressA: %s\nAddressB: %s\n", addressA, addressB)
+	case bytes.HasPrefix(kvA.Key, types.DTagTransferRequestPrefix):
```
Comment (RiccardoM) [inline x/profiles/simulation/decoder.go:26]: Since it looks like the `DTagTransferRequestPrefix` will represent a list of transfer requests, shouldn't it be more appropriate to rename it `DTagTransferRequestsPrefix`?

File: x/profiles/simulation/operations_profile.go [hunk 9]
```diff
@@ -29,13 +29,13 @@ func SimulateMsgSaveProfile(k keeper.Keeper, ak auth.AccountKeeper) sim.Operatio
 			return sim.NoOpMsg(types.ModuleName), nil, nil
 		}
 
-		var profilePic, coverPic = "", ""
+		var profilePic, coverPic = RandomProfilePic(r), RandomProfileCover(r)
```
Comment (RiccardoM) [inline x/profiles/simulation/operations_profile.go:32]: Why was this changed? Shouldn't we also test if the profile pic and cover pics are empty?

File: x/profiles/types/events.go [hunk 10]
```diff
@@ -1,11 +1,18 @@
 package types
 
 const (
-	EventTypeProfileSaved   = "profile_saved"
-	EventTypeProfileDeleted = "profile_deleted"
+	EventTypeProfileSaved            = "profile_saved"
+	EventTypeProfileDeleted          = "profile_deleted"
+	EventTypeDTagTransferRequest     = "dtag_transfer_request"
+	EventTypeDTagTransferReqAccepted = "accepted_transfer_request"
```
Comment (RiccardoM) [inline x/profiles/types/events.go:7]: I think it would be more appropriate to rename this event to 

```suggestion
	EventTypeDTagTransferAccept = "dtag_transfer_accept"
```

To make it more coherent wit the request one

File: x/profiles/types/events.go [hunk 11]
```diff
@@ -1,11 +1,18 @@
...
 
 	// Profile attributes
 	AttributeProfileDtag         = "profile_dtag"
 	AttributeProfileCreator      = "profile_creator"
 	AttributeProfileCreationTime = "profile_creation_time"
+
+	// DTag trade attributes
+	AttributeCurrentOwner  = "current_owner"
+	AttributeReceivingUser = "receiving_user"
+	AttributeNewDTag       = "new_DTag"
```
Comment (RiccardoM) [inline x/profiles/types/events.go:17]: I think it's better to make this all lowercase

```suggestion
	AttributeNewDTag       = "new_dtag"
```

File: x/posts/types/models/post_response_test.go [hunk 12]
```diff
@@ -175,7 +175,7 @@ func TestPostQueryResponse_String(t *testing.T) {
 		},
 	)
 
-	expected := "ID: 93ed93d2f3b3363399c8b7a4509f0530a8c863c16c755bf916901ab55ce33322\nReactions: [[Shortcode] :like: [Value] https://example.com/like [Owner] cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4 [Shortcode] :+1: [Value] 👍 [Owner] cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae]\nChildren: [dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1, dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1]"
+	expected := "ID: 93ed93d2f3b3363399c8b7a4509f0530a8c863c16c755bf916901ab55ce33322\nReactions: [[Shortcode] :like: [Value] https://example.com/like [CurrentOwner] cosmos1s3nh6tafl4amaxkke9kdejhp09lk93g9ev39r4 [Shortcode] :+1: [Value] 👍 [CurrentOwner] cosmos15lt0mflt6j9a9auj7yl3p20xec4xvljge0zhae]\nChildren: [dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1, dd065b70feb810a8c6f535cf670fe6e3534085221fa964ed2660ebca93f910d1]"
```
Comment (RiccardoM) [inline x/posts/types/models/post_response_test.go:178]: Why was `Owner` changed to `CurrentOwner`?

File: x/posts/types/models/reactions/post_reaction.go [hunk 13]
```diff
@@ -32,7 +32,7 @@ func NewPostReaction(shortcode, value string, owner sdk.AccAddress) PostReaction
 
 // String implements fmt.Stringer
 func (reaction PostReaction) String() string {
-	out := fmt.Sprintf("[Shortcode] %s [Value] %s [Owner] %s", reaction.Shortcode, reaction.Value, reaction.Owner)
+	out := fmt.Sprintf("[Shortcode] %s [Value] %s [CurrentOwner] %s", reaction.Shortcode, reaction.Value, reaction.Owner)
```
Comment (RiccardoM) [inline x/posts/types/models/reactions/post_reaction.go:35]: I think it's better to leave this as `Owner` instead. I don't see how a post reaction could change owner over time. 

```suggestion
	out := fmt.Sprintf("[Shortcode] %s [Value] %s [Owner] %s", reaction.Shortcode, reaction.Value, reaction.Owner)
```

File: x/profiles/types/msgs/msgs.go [hunk 14]
```diff
@@ -104,3 +104,109 @@ func (msg MsgDeleteProfile) GetSignBytes() []byte {
...
+	return []sdk.AccAddress{msg.ReceivingUser}
+}
+
+// ----------------------
+// --- MsgAcceptDTagTransfer
+// ----------------------
+
+// MsgAcceptDTagTransfer represent a DTag transfer acceptance message
+type MsgAcceptDTagTransfer struct {
+	NewDTag       string         `json:"new_d_tag" yaml:"new_d_tag"`
```
Comment (RiccardoM) [inline x/profiles/types/msgs/msgs.go:164]: I think it's better to rename this `new_dtag` instead in order to be coherent with the namings all over

```suggestion
	NewDTag       string         `json:"new_d_tag" yaml:"new_dtag"`
```

File: x/reports/types/models/report.go [hunk 15]
```diff
@@ -44,7 +44,7 @@ type Reports []Report
 
 // String implements stringer
 func (reports Reports) String() string {
-	out := "Type - Message - Sender\n"
+	out := "Type - Message - ReceivingUser\n"
```
Comment (RiccardoM) [inline x/reports/types/models/report.go:47]: I think it's better to leave it as it was

```suggestion
	out := "Type - Message - Sender\n"
```

**Review by RiccardoM — COMMENTED** [review]
I've suggested renaming some `dTag` to `DTag`. Can you please also apply this renaming to all the left usages? I think it's better to always use `DTag` instead of `dTag` and `DTag` for coherence.

File: x/posts/types/models/reactions/post_reaction.go (hunk 13 above)
Comment (bragaz) [reply x/posts/types/models/reactions/post_reaction.go:35]: refactoring error

File: x/profiles/simulation/operations_profile.go (hunk 9 above)
Comment (bragaz) [reply x/profiles/simulation/operations_profile.go:32]: A particular seed raised a bug inside this test.
Basically if there was no pictures it would have assigned the empty string leading to an error when later inside
the profile validation.
You were right to point me this because I found a better solution which is set the two vars to `nil` by default instead of an empty string, which is basically what happens if someone doesn't share any profile image.

File: docs/developers/msgs/profiles/accept-dtag-transfer-request.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
+# `MsgAcceptDTagTransfer`
+This message allows you to accept a `DTag`'s transfer request.
+
+## Structure
+````json
+{
+  "type": "desmos/MsgAcceptDTagTransfer",
+  "value": {
+    "new_d_tag": "<The new DTag for the current owner's profile>",
```
Comment (RiccardoM) [inline docs/developers/msgs/profiles/accept-dtag-transfer-request.md:9]:
```suggestion
    "new_dtag": "<The new DTag for the current owner's profile>",
```

File: docs/developers/msgs/profiles/accept-dtag-transfer-request.md [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    "current_owner": "<Desmos address of the DTag owner>",
+    "receiving_user": "<Desmos address that's making the DTag's request>"
+  }
+}
+````
+
+### Attributes
+| Attribute | Type | Description |
+| :-------: | :----: | :-------- |
+| `new_d_tag` | String | The new `DTag` for the current owner profile that will replace the traded one |
```
Comment (RiccardoM) [inline docs/developers/msgs/profiles/accept-dtag-transfer-request.md:19]:
```suggestion
| `new_dtag` | String | The new `DTag` for the current owner profile that will replace the traded one |
```

File: docs/developers/msgs/profiles/accept-dtag-transfer-request.md [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+| `current_owner`  | String | Desmos address of the user that is the owner of the requested `DTag` |
+| `receiving_user`| String | Desmos address of the user that request the `DTag` |
+
+## Example
+````json
+{
+  "type": "desmos/MsgAcceptDTagTransfer",
+  "value": {
+    "new_d_tag": "newDTag",
```
Comment (RiccardoM) [inline docs/developers/msgs/profiles/accept-dtag-transfer-request.md:28]:
```suggestion
    "new_dtag": "newDTag",
```

File: docs/developers/query-data.md [hunk 19]
```diff
@@ -14,6 +14,7 @@ Aside from [performing transactions](perform-transactions.md) you can also query
 ## Profiles
 - [Query a profile](queries/profile.md)
 - [Query the stored profiles](queries/profiles.md)
+- [Query all the DTag transfer's requests](queries/user-dtag-requests.md)
```
Comment (RiccardoM) [inline docs/developers/query-data.md:17]:
```suggestion
- [Query all the DTag transfer requests](queries/user-dtag-requests.md)
```

File: x/profiles/keeper/querier_test.go [hunk 20]
```diff
@@ -172,3 +172,61 @@ func (suite *KeeperTestSuite) Test_queryParams() {
...
+			name: "Stored dTag requests returns correctly",
+			path: []string{types.QueryDTagRequests, suite.testData.user.String()},
+			storedRequests: []types.DTagTransferRequest{
+				types.NewDTagTransferRequest(suite.testData.user, suite.testData.otherUser),
+			},
+			expResult: []types.DTagTransferRequest{
+				types.NewDTagTransferRequest(suite.testData.user, suite.testData.otherUser),
+			},
+			expErr: nil,
+		},
```
Comment (RiccardoM) [inline x/profiles/keeper/querier_test.go:201]: Can we add a test to make sure that querying an empty list works properly?

File: x/profiles/types/genesis.go [hunk 21]
```diff
@@ -1,33 +1,27 @@
...
-)
-
 // GenesisState contains the data of the genesis state for the profile module
 type GenesisState struct {
-	Profiles           []Profile                   `json:"profiles" yaml:"profiles"`
-	Params             Params                      `json:"params" yaml:"params"`
-	UsersRelationships map[string][]sdk.AccAddress `json:"users_relationships"`
+	Profiles             []Profile             `json:"profiles" yaml:"profiles"`
+	Params               Params                `json:"params" yaml:"params"`
+	DTagTransferRequests []DTagTransferRequest `json:"dtag_transfer_requests" yaml:"dtag_transfer_requests"`
```
Comment (RiccardoM) [inline x/profiles/types/genesis.go:7]: Why was the `UsersRelationship` deleted from the genesis file?

**Review by RiccardoM — COMMENTED** [review]
Some things should be addressed. 

Also, I think that when users **change** their DTag, all the pending transfer requests should be deleted. This is to prevent scammers that: 

1. Receive a DTag transfer request
2. Change their DTag to something else
3. Accept the transfer request
4. Get their DTag back

File: x/profiles/types/genesis.go (hunk 21 above)
Comment (bragaz) [reply x/profiles/types/genesis.go:7]: That field was an unused one that was left there forgotten after I moved `relationships` into a standalone module.

File: x/profiles/client/cli/tx.go [hunk 22]
```diff
@@ -104,3 +106,55 @@ func GetCmdDeleteProfile(cdc *codec.Codec) *cobra.Command {
 
 	return cmd
 }
+
+// GetCmdRequestDTagTransfer is the CLI command for request an existent dTag to the given owner
+func GetCmdRequestDTagTransfer(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "transfer-dtag [address]",
+		Short: "Make a request to get the dTag of the given address Profile",
```
Comment (RiccardoM) [inline x/profiles/client/cli/tx.go:114]:
```suggestion
		Short: "Make a request to get the DTag of the user having the given address",
```

File: x/profiles/client/cli/tx.go [hunk 23]
```diff
@@ -104,3 +106,55 @@ func GetCmdDeleteProfile(cdc *codec.Codec) *cobra.Command {
...
+
+	return cmd
+}
+
+// GetCmdAcceptDTagTransfer is the CLI command to accept the request to transfer a dTag
+func GetCmdAcceptDTagTransfer(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "accept-dtag-transfer.md [newDTag] [address]",
+		Short: "Accept a dtag transfer request made by the user with the given address",
+		Long: `Accept a previously made dtag transfer request from the given address.
```
Comment (RiccardoM) [inline x/profiles/client/cli/tx.go:140]:
```suggestion
		Use:   "accept-dtag-transfer [new DTag] [address]",
		Short: "Accept a DTag transfer request made by the user with the given address",
```

File: x/profiles/client/rest/query.go [hunk 24]
```diff
@@ -15,6 +15,23 @@ func registerQueryRoutes(cliCtx context.CLIContext, r *mux.Router) {
 	r.HandleFunc("/profiles/parameters", queryProfilesParamsHandlerFn(cliCtx)).Methods("GET")
 	r.HandleFunc("/profiles/{address_or_dtag}", queryProfileHandlerFn(cliCtx)).Methods("GET")
 	r.HandleFunc("/profiles", queryProfilesHandlerFn(cliCtx)).Methods("GET")
+	r.HandleFunc("/dtag/requests/{address}", queryDTagRequests(cliCtx)).Methods("GET")
```
Comment (RiccardoM) [inline x/profiles/client/rest/query.go:18]: I personally think it's better to change this endpoint making it more similar to the `profile` one: 

```suggestion
	r.HandleFunc("/profiles/{address}/dtag-transfer-requests", queryDTagRequests(cliCtx)).Methods("GET")
```

File: x/profiles/client/rest/tx.go [hunk 25]
```diff
@@ -14,6 +14,8 @@ import (
 func registerTxRoutes(cliCtx context.CLIContext, r *mux.Router) {
 	r.HandleFunc("/profiles/{address}", saveProfileHandler(cliCtx)).Methods("PUT")
 	r.HandleFunc("/profiles/{address}", deleteProfileHandler(cliCtx)).Methods("DELETE")
+	r.HandleFunc("/transfer/dtag", requestDTagTransferHandler(cliCtx)).Methods("POST")
+	r.HandleFunc("/accept/dtag/transfer", acceptTransferRequestHandler(cliCtx)).Methods("POST")
```
Comment (RiccardoM) [inline x/profiles/client/rest/tx.go:18]: I think it's probably better to make those two endpoints more similar to one another

```suggestion
	r.HandleFunc("/dtag-transfer-requests", requestDTagTransferHandler(cliCtx)).Methods("POST")
	r.HandleFunc("/dtag-transfer-requests/accept", acceptTransferRequestHandler(cliCtx)).Methods("POST")
```

File: x/profiles/keeper/handler.go [hunk 26]
```diff
@@ -82,6 +86,11 @@ func handleMsgSaveProfile(ctx sdk.Context, keeper Keeper, msg types.MsgSaveProfi
 		profile = types.NewProfile(msg.Dtag, msg.Creator, ctx.BlockTime())
 	}
 
+	// If dTag changes, delete all the previous dTag transfer requests
```
Comment (RiccardoM) [inline x/profiles/keeper/handler.go:89]:
```suggestion
	// If the DTag changes, delete all the previous DTag transfer requests
```

File: x/profiles/keeper/handler.go [hunk 27]
```diff
@@ -140,3 +149,88 @@ func handleMsgDeleteProfile(ctx sdk.Context, keeper Keeper, msg types.MsgDeleteP
...
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, fmt.Sprintf("no request made from %s", msg.ReceivingUser))
+	}
+
+	// Get the current owner profile
+	currentOwnerProfile, exist := keeper.GetProfile(ctx, msg.CurrentOwner)
+	if !exist {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, fmt.Sprintf("profile of %s doesn't exist", msg.CurrentOwner))
+	}
+
+	// Get the dTag to trade
```
Comment (RiccardoM) [inline x/profiles/keeper/handler.go:198]:
```suggestion
	// Get the DTag to trade
```

File: x/profiles/keeper/handler.go [hunk 28]
```diff
@@ -140,3 +149,88 @@ func handleMsgDeleteProfile(ctx sdk.Context, keeper Keeper, msg types.MsgDeleteP
...
+	} else {
+		receiverProfile = receiverProfile.WithDTag(dTagToTrade)
+	}
+
+	err = keeper.SaveProfile(ctx, receiverProfile)
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
+	}
+
+	keeper.DeleteAllDTagTransferRequests(ctx, msg.CurrentOwner)
```
Comment (RiccardoM) [inline x/profiles/keeper/handler.go:221]: I think it's better to delete also all the DTag transfer requests associated with the `ReceivingUser` to make sure he/she doesn't fraud anyone

File: x/profiles/keeper/handler.go [hunk 29]
```diff
@@ -140,3 +149,88 @@ func handleMsgDeleteProfile(ctx sdk.Context, keeper Keeper, msg types.MsgDeleteP
...
+	transferRequest := types.NewDTagTransferRequest(msg.CurrentOwner, msg.ReceivingUser)
+
+	if err := keeper.SaveDTagTransferRequest(ctx, transferRequest); err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
+	}
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeDTagTransferRequest,
+		sdk.NewAttribute(types.AttributeCurrentOwner, transferRequest.CurrentOwner.String()),
+		sdk.NewAttribute(types.AttributeReceivingUser, transferRequest.ReceivingUser.String()),
```
Comment (RiccardoM) [inline x/profiles/keeper/handler.go:164]: Proably we can add an attribute to tell which DTag the user wants to trade?

File: x/profiles/keeper/handler.go [hunk 30]
```diff
@@ -140,3 +149,88 @@ func handleMsgDeleteProfile(ctx sdk.Context, keeper Keeper, msg types.MsgDeleteP
 
 	return &result, nil
 }
+
+// handleMsgRequestDTagTransfer handles the request of a dTag transfer
+func handleMsgRequestDTagTransfer(ctx sdk.Context, keeper Keeper, msg types.MsgRequestDTagTransfer) (*sdk.Result, error) {
+	transferRequest := types.NewDTagTransferRequest(msg.CurrentOwner, msg.ReceivingUser)
```
Comment (RiccardoM) [inline x/profiles/keeper/handler.go:155]: What if we also save the DTag that is the object of the trade? This way we can later check to make sure it hasn't been edited.

File: x/profiles/keeper/handler.go [hunk 31]
```diff
@@ -140,3 +149,88 @@ func handleMsgDeleteProfile(ctx sdk.Context, keeper Keeper, msg types.MsgDeleteP
...
+	if err != nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, err.Error())
+	}
+
+	keeper.DeleteAllDTagTransferRequests(ctx, msg.CurrentOwner)
+
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeDTagTransferAccept,
+		sdk.NewAttribute(types.AttributeCurrentOwner, msg.CurrentOwner.String()),
+		sdk.NewAttribute(types.AttributeReceivingUser, msg.ReceivingUser.String()),
```
Comment (RiccardoM) [inline x/profiles/keeper/handler.go:226]: I think we can add a couple of attributes here like 

- the traded DTag
- the new DTag that the giver has chosen

File: x/profiles/keeper/keeper.go [hunk 32]
```diff
@@ -136,3 +136,59 @@ func (k Keeper) GetProfile(ctx sdk.Context, address sdk.AccAddress) (profile typ
...
+// SaveDTagTransferRequest save the given request into the currentOwner's requests
+// returning errors if an equal one already exists.
+func (k Keeper) SaveDTagTransferRequest(ctx sdk.Context, transferRequest types.DTagTransferRequest) error {
+	store := ctx.KVStore(k.StoreKey)
+	key := types.DtagTransferRequestStoreKey(transferRequest.CurrentOwner)
+
+	var requests []types.DTagTransferRequest
+	bz := store.Get(key)
+	if bz != nil {
+		k.Cdc.MustUnmarshalBinaryBare(bz, &requests)
```
Comment (RiccardoM) [inline x/profiles/keeper/keeper.go:149]: Is this check really necessary? Couldn't it simply be...?

```go
k.Cdc.MustUnmarshalBinaryBare(store.Get(key), &requests)
```

File: x/profiles/keeper/keeper_test.go [hunk 33]
```diff
@@ -216,3 +215,162 @@ func (suite *KeeperTestSuite) TestKeeper_GetProfiles() {
...
+				suite.testData.user, suite.testData.otherUser)},
+		},
+		{
+			name:               "not already present request saved correctly",
+			storedTransferReqs: nil,
+			transferReq:        types.NewDTagTransferRequest(suite.testData.user, suite.testData.otherUser),
+			expErr:             nil,
+			expStoredTransferReqs: []types.DTagTransferRequest{types.NewDTagTransferRequest(
+				suite.testData.user, suite.testData.otherUser)},
+		},
```
Comment (RiccardoM) [inline x/profiles/keeper/keeper_test.go:245]: Can you add some tests with an already existing request? Particularly: 

- storing a new request with a different current owner address works 
- storing a new request with a different receiver address works
- storing a new request with a different DTag works

**Review by RiccardoM — COMMENTED** [review]
Looks good to me overall. Some small changes but then it should be ready to be merged

File: x/profiles/keeper/handler.go (hunk 29 above)
Comment (bragaz) [reply x/profiles/keeper/handler.go:164]: This means using `keeper.GetDtagFromAddress(ctx, address)` function just to retrieve the DTag.
Is it worth it?

**Review by RiccardoM — APPROVED** [review]
